keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,752 @@
+   nodes, and pod anti-affinity forbids to run two pods on the same node.
+3. Terminating pods can still do useful work or hold old connections. Users would like to track
+   this work through the deployment's status. See
+   [kubernetes/kubernetes#110171](https://github.com/kubernetes/kubernetes/issues/110171) for more
+   details.
+
+[kubernetes/kubernetes#107920](https://github.com/kubernetes/kubernetes/issues/107920) issue is covering this as well.
+
+### Goals
+
+- Deployments should allow an option to either wait for its pods to terminate before creating new
+  pods, or to create the pods immediately. This should take into consideration the Deployment
+  strategy.
+- Deployments and ReplicaSets should indicate a number of managed terminating pods in their status
+  field.
+
+### Non-Goals
+
+## Proposal
+
+This KEP proposes to introduce a new  `.spec.podReplacementPolicy` field (similar to Job's
+`.spec.podReplacementPolicy` in [kubernetes/enhancements#3939](https://github.com/kubernetes/enhancements/issues/3939))
+that would control how many pods should be present at any given time.
+
+The termination of a Deployment/ReplicaSet pod is always triggered by a pod deletion due to
+an enforced pod field `restartPolicy: Always`.
+
+We are distinguishing between terminating and terminated pods. 
+- Terminating pods are running pods with a `deletionTimestamp`. 
+- Terminated pods are pods with a `deletionTimestamp` that have reached the `Succeeded` or `Failed` phase
+  and are subsequently removed from etcd.
+
+Unfortunately, the current behavior is inconsistent with how we treat terminating and terminated pods
+in the deployment controller.
+
+- The Recreate Deployment strategy waits for terminating pods to terminate before creating
+  (scheduling) new pods.
+- The RollingUpdate deployment strategy does not wait for terminating pods and creates (schedules)
+  new pods immediately.
+- Scaling up a Deployment also does not wait for terminating pods and creates (schedules) new pods
+  right away.
+
+Unfortunately, in Deployments with a Recreate strategy we can get mixed behavior. The
+deployment will wait for old pods to terminate during a rollout, but will ignore the terminating
+pods when scaling the pods. So it is still possible to end up with a larger number of pods than
+`.spec.replicas`.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an application user, I would prefer predictable number of pods in my cluster to prevent any
+scheduling issues and unnecessary autoscaling of nodes. I would also like to achieve consistent
+allocation of other scarce resources to pods.
+
+#### Story 2
+
+As an application user, I would like to keep the old behavior of fast scaling of pods and do not
+mind the higher utilization of resources.
+
+#### Story 3
+
+As an application user, I would like to track the number of instances that perform useful work
+during the entire lifecycle of a pod.
+
+### Notes/Constraints/Caveats (Optional)
```

Comment:
Could you maybe add why this is only targeted for Deployment? Just so its documented elsewhere.

Reply:
+1